File: src/flask/app/api/routes/scheduler.py
"""Routes for scheduler assignment management."""

from datetime import datetime, timezone
from typing import Any

from pydantic import BaseModel
from flask import Blueprint, jsonify, request

from app.models import TimeOffRequestPublic

from app.api.deps import (
    get_current_user,
    get_current_active_superuser,
    get_db,
    require_scope,
)
from app.models import (
    AssignmentPublic,
    AssignmentsPublic,
    TimeOffRequest,
    TimeOffRequestStatus,
    User,
)

from app.requests.assignment_request import (
    AssignmentCreate,
    AssignmentUpdate,
    BulkAssignRequest,
    TimeOffRequestCreate,
)
from app.repositories.assignment_repo import AssignmentRepository
from app.services.scheduler_service import SchedulerService
from sqlalchemy import select

router = Blueprint("scheduler", __name__, url_prefix="/scheduler")


class TimeOffRequestsPublic(BaseModel):
    data: list[TimeOffRequestPublic]
    count: int


class EnrichedAssignment(AssignmentPublic):
    user_email: str
    user_full_name: str | None


class EnrichedAssignmentsResponse(BaseModel):
    data: list[EnrichedAssignment]
    count: int


# ── Time-off request routes ──────────


@router.route("/time-off-requests", methods=["GET"])
@require_scope("member:limited")
def get_my_time_off(start_date: str = None, end_date: str = None):
    session = get_db()
    current_user = get_current_user()
    stmt = select(TimeOffRequest).where(TimeOffRequest.user_id == current_user.id)
    if start_date and end_date:
        start = datetime.fromisoformat(start_date)
        end = datetime.fromisoformat(end_date)
        stmt = stmt.where(TimeOffRequest.date >= start).where(TimeOffRequest.date <= end)
    stmt = stmt.order_by(TimeOffRequest.date)
    result = session.execute(stmt)
    items = result.scalars().all()
    return jsonify(
        TimeOffRequestsPublic(
            data=[TimeOffRequestPublic.model_validate(a) for a in items], count=len(items)
        ).model_dump()
    )


@router.route("/time-off-request", methods=["POST"])
@require_scope("member:limited")
def create_time_off_request():
    session = get_db()
    current_user = get_current_user()
    data = request.get_json()
    data_request = TimeOffRequestCreate(**data)
    row = TimeOffRequest(
        user_id=current_user.id,
        date=data_request.date,
        status=TimeOffRequestStatus.pending,
        notes=data_request.notes,
        created_on=datetime.now(timezone.utc),
        updated_on=None,
    )
    session.add(row)
    session.commit()
    session.refresh(row)
    return jsonify(TimeOffRequestPublic.model_validate(row).model_dump()), 201


@router.route("/time-off-requests/<time_off_id>/approve", methods=["PATCH"])
@require_scope("scheduler:admin")
def approve_time_off_request(time_off_id: str):
    session = get_db()
    _ = get_current_active_superuser()
    row = session.get(TimeOffRequest, time_off_id)
    if not row:
        return jsonify({"detail": "Time-off request not found"}), 404
    row.status = TimeOffRequestStatus.approved
    row.updated_on = datetime.now(timezone.utc)
    session.add(row)
    session.commit()
    session.refresh(row)
    return jsonify(TimeOffRequestPublic.model_validate(row).model_dump())


@router.route("/time-off-requests/<time_off_id>/decline", methods=["PATCH"])
@require_scope("scheduler:admin")
def decline_time_off_request(time_off_id: str):
    session = get_db()
    _ = get_current_active_superuser()
    row = session.get(TimeOffRequest, time_off_id)
    if not row:
        return jsonify({"detail": "Time-off request not found"}), 404
    row.status = TimeOffRequestStatus.declined
    row.updated_on = datetime.now(timezone.utc)
    session.add(row)
    session.commit()
    session.refresh(row)
    return jsonify(TimeOffRequestPublic.model_validate(row).model_dump())


@router.route("/time-off-requests/<time_off_id>", methods=["DELETE"])
@require_scope("member:limited")
def delete_time_off_request(time_off_id: str):
    session = get_db()
    current_user = get_current_user()
    row = session.get(TimeOffRequest, time_off_id)
    if not row:
        return jsonify({"detail": "Time-off request not found"}), 404
    if row.user_id != current_user.id:
        return jsonify({"detail": "Can only delete your own time-off requests"}), 403
    session.delete(row)
    session.commit()
    return jsonify({}), 204


# ── Assignment routes ────


@router.route("/my-assignments", methods=["GET"])
@require_scope("member:limited")
def get_my_assignments():
    session = get_db()
    current_user = get_current_user()
    repo = AssignmentRepository(session=session)
    assignments = repo.get_by_user_id(current_user.id)
    return jsonify(
        AssignmentsPublic(
            data=[AssignmentPublic.model_validate(a) for a in assignments], count=len(assignments)
        ).model_dump()
    )


@router.route("/calendar", methods=["GET"])
@require_scope("member:limited")
def get_calendar_assignments(start_date: str = None, end_date: str = None):
    session = get_db()
    _ = get_current_user()
    start = datetime.fromisoformat(start_date)
    end = datetime.fromisoformat(end_date)
    repo = AssignmentRepository(session=session)
    assignments = repo.get_by_date_range(start, end)
    return jsonify(
        AssignmentsPublic(
            data=[AssignmentPublic.model_validate(a) for a in assignments], count=len(assignments)
        ).model_dump()
    )


@router.route("/calendar-with-names", methods=["GET"])
@require_scope("member:limited")
def get_calendar_with_names(start_date: str = None, end_date: str = None):
    session = get_db()
    _ = get_current_user()
    start = datetime.fromisoformat(start_date)
    end = datetime.fromisoformat(end_date)
    repo = AssignmentRepository(session=session)
    assignments = repo.get_by_date_range(start, end)
    enriched: list[EnrichedAssignment] = []
    for a in assignments:
        user = session.get(User, a.user_id)
        enriched.append(
            EnrichedAssignment(
                **AssignmentPublic.model_validate(a).model_dump(),
                user_email=user.email if user else "unknown",
                user_full_name=(
                    user.full_name
                    if user and user.full_name
                    else (user.email if user else "unknown")
                ),
            )
        )
    return jsonify(EnrichedAssignmentsResponse(data=enriched, count=len(enriched)).model_dump())


@router.route("/my-calendar", methods=["GET"])
@require_scope("member:limited")
def get_my_calendar(start_date: str = None, end_date: str = None):
    session = get_db()
    current_user = get_current_user()
    start = datetime.fromisoformat(start_date)
    end = datetime.fromisoformat(end_date)
    repo = AssignmentRepository(session=session)
    assignments = repo.get_by_user_and_date_range(current_user.id, start, end)
    return jsonify(
        AssignmentsPublic(
            data=[AssignmentPublic.model_validate(a) for a in assignments], count=len(assignments)
        ).model_dump()
    )


@router.route("/", methods=["GET"])
@require_scope("scheduler:admin")
def list_assignments(skip: int = 0, limit: int = 100):
    session = get_db()
    _ = get_current_active_superuser()
    repo = AssignmentRepository(session=session)
    assignments, total_count = repo.get_all(skip=skip, limit=limit)
    return jsonify(
        AssignmentsPublic(
            data=[AssignmentPublic.model_validate(a) for a in assignments], count=total_count
        ).model_dump()
    )


@router.route("/<assignment_id>", methods=["GET"])
@require_scope("scheduler:admin")
def get_assignment(assignment_id: str):
    session = get_db()
    _ = get_current_active_superuser()
    repo = AssignmentRepository(session=session)
    assignment = repo.get_by_id(assignment_id)
    if not assignment:
        return jsonify({"detail": "Assignment not found"}), 404
    return jsonify(AssignmentPublic.model_validate(assignment).model_dump())


@router.route("/", methods=["POST"])
@require_scope("scheduler:admin")
def create_assignment():
    session = get_db()
    _ = get_current_active_superuser()
    data = request.get_json()
    assignment_in = AssignmentCreate(**data)
    repo = AssignmentRepository(session=session)
    conflicts = repo.check_conflicts(
        user_id=assignment_in.user_id, event_date=assignment_in.event_date
    )
    if conflicts:
        conflict_ids = [c.id for c in conflicts]
        conflict_details = "\n".join(
            f"  - {c.type}: {c.role} on {c.event_date.strftime('%Y-%m-%d')}" for c in conflicts
        )
        return (
            jsonify(
                {
                    "message": "User already has an assignment on this date.",
                    "conflicts": [c.model_dump() for c in conflicts],
                    "conflict_ids": conflict_ids,
                    "details": conflict_details,
                }
            ),
            409,
        )
    assignment = repo.create(assignment_in=assignment_in)
    if assignment is None:
        return jsonify({"detail": "Failed to create assignment"}), 500
    scheduler_svc = SchedulerService(session)
    scheduler_svc.send_assignment_notification(
        user_id=assignment.user_id,
        assignment_type=assignment_in.type.value,
        role=assignment_in.role,
        event_date=assignment.event_date.strftime("%B %d, %Y"),
        instrument=assignment_in.instrument,
        notes=assignment_in.notes,
    )
    return jsonify(AssignmentPublic.model_validate(assignment).model_dump()), 201


@router.route("/<assignment_id>", methods=["PATCH"])
@require_scope("scheduler:admin")
def update_assignment(assignment_id: str):
    session = get_db()
    _ = get_current_active_superuser()
    data = request.get_json()
    assignment_in = AssignmentUpdate(**data)
    repo = AssignmentRepository(session=session)
    assignment = repo.get_by_id(assignment_id)
    if not assignment:
        return jsonify({"detail": "Assignment not found"}), 404
    assignment = repo.update(db_assignment=assignment, assignment_in=assignment_in)
    return jsonify(AssignmentPublic.model_validate(assignment).model_dump())


@router.route("/<assignment_id>", methods=["DELETE"])
@require_scope("scheduler:admin")
def delete_assignment(assignment_id: str):
    session = get_db()
    _ = get_current_active_superuser()
    repo = AssignmentRepository(session=session)
    assignment = repo.get_by_id(assignment_id)
    if not assignment:
        return jsonify({"detail": "Assignment not found"}), 404
    repo.delete(db_assignment=assignment)
    return jsonify({}), 204


class BulkAssignResponse(BaseModel):
    created: list[AssignmentPublic]
    conflicts: list[dict[str, Any]]


@router.route("/bulk", methods=["POST"])
@require_scope("scheduler:admin")
def bulk_assign():
    session = get_db()
    _ = get_current_active_superuser()
    data = request.get_json()
    bulk_in = BulkAssignRequest(**data)
    repo = AssignmentRepository(session=session)
    scheduler_svc = SchedulerService(session)
    created: list[AssignmentPublic] = []
    conflicts: list[dict[str, Any]] = []
    for entry in bulk_in.entries:
        assignment_in = AssignmentCreate(
            user_id=entry.user_id,
            event_date=bulk_in.event_date,
            type=bulk_in.type,
            role=entry.role,
            instrument=entry.instrument,
            notes=entry.notes,
            group_leader=entry.group_leader,
        )
        conflicts_check = repo.check_conflicts(user_id=entry.user_id, event_date=bulk_in.event_date)
        if conflicts_check:
            conflicts.append(
                {
                    "user_id": entry.user_id,
                    "message": "User already has an assignment on "
                    + bulk_in.event_date.strftime("%Y-%m-%d"),
                    "conflicts": [c.model_dump() for c in conflicts_check],
                }
            )
            continue
        assignment = repo.create(assignment_in=assignment_in)
        if assignment is None:
            conflicts.append({"user_id": entry.user_id, "message": "Failed to create assignment"})
            continue
        scheduler_svc.send_assignment_notification(
            user_id=assignment.user_id,
            assignment_type=bulk_in.type.value,
            role=entry.role,
            event_date=assignment.event_date.strftime("%B %d, %Y"),
            instrument=entry.instrument,
            notes=entry.notes,
        )
        created.append(AssignmentPublic.model_validate(assignment))
    return jsonify(BulkAssignResponse(created=created, conflicts=conflicts).model_dump()), 201

File: src/fe/lib/api/scheduler.ts
const API_BASE = process.env.NEXT_PUBLIC_API_URL || "https://localhost:8000/";
const API_V1 = "api/v1";

import { fetchWithAuth } from "./auth";

// --- Scheduler / Assignment types ---

export interface Assignment {
  id: string;
  user_id: string;
  user_email: string;
  user_full_name: string | null;
  event_date: string;
  type: "music" | "service";
  role: string;
  instrument: string | null;
  notes: string | null;
  group_leader: boolean;
  created_on: string;
  updated_on: string | null;
}

export interface AssignmentConflict {
  id: string;
  type: string;
  role: string;
  event_date: string;
}

export interface AssignmentCreateInput {
  user_id: string;
  event_date: string;
  type: "music" | "service";
  role?: string;
  instrument?: string | null;
  notes?: string | null;
}

export interface AssignmentUpdateInput {
  event_date?: string;
  type?: "music" | "service";
  role?: string;
  instrument?: string | null;
  notes?: string | null;
  user_id?: string;
}

export interface BulkAssignEntry {
  user_id: string;
  role: string;
  instrument: string | null;
  notes: string | null;
  group_leader: boolean;
}

export interface BulkAssignRequest {
  event_date: string;
  type: "music" | "service";
  entries: BulkAssignEntry[];
}

export interface BulkAssignConflict {
  user_id: string;
  message: string;
  conflicts?: AssignmentConflict[];
}

export interface BulkAssignResponse {
  created: Assignment[];
  conflicts: BulkAssignConflict[];
}

export interface AssignmentsResponse {
  data: Assignment[];
  count: number;
}

// --- Time-off request types ---

export interface TimeOffRequest {
  id: string;
  user_id: string;
  date: string;
  status: "pending" | "approved" | "declined";
  notes: string | null;
  created_on: string;
  updated_on: string | null;
}

export interface TimeOffRequestsResponse {
  data: TimeOffRequest[];
  count: number;
}

// --- Scheduler API functions ---

export async function fetchAssignments(): Promise<AssignmentsResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/`);
  if (!res.ok) {
    throw new Error("Failed to fetch assignments");
  }
  return res.json();
}

export async function fetchAssignment(id: string): Promise<Assignment> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/${id}`);
  if (!res.ok) {
    throw new Error("Failed to fetch assignment");
  }
  return res.json();
}

export async function createAssignment(data: AssignmentCreateInput): Promise<Assignment> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    if (res.status === 409) {
      const detail = JSON.parse(body);
      throw new Error(detail.detail?.message || detail.detail || "Conflict: user already has an assignment on this date.");
    }
    throw new Error(body || "Failed to create assignment");
  }
  return res.json();
}

export async function updateAssignment(id: string, data: AssignmentUpdateInput): Promise<Assignment> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/${id}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to update assignment");
  }
  return res.json();
}

export async function deleteAssignment(id: string): Promise<void> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/${id}`, {
    method: "DELETE",
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to delete assignment");
  }
}

export async function bulkAssignAssignments(data: BulkAssignRequest): Promise<BulkAssignResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/bulk`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to bulk assign assignments");
  }
  return res.json();
}

export async function fetchMyAssignments(): Promise<AssignmentsResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/my-assignments`);
  if (!res.ok) {
    throw new Error("Failed to fetch my assignments");
  }
  return res.json();
}

export async function fetchCalendarAssignments(startDate: string, endDate: string): Promise<AssignmentsResponse> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/calendar?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`
  );
  if (!res.ok) {
    throw new Error("Failed to fetch calendar assignments");
  }
  return res.json();
}

export async function fetchCalendarWithNames(startDate: string, endDate: string): Promise<AssignmentsResponse> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/calendar-with-names?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`
  );
  if (!res.ok) {
    throw new Error("Failed to fetch calendar with names");
  }
  return res.json();
}

export async function fetchMyCalendar(startDate: string, endDate: string): Promise<AssignmentsResponse> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/my-calendar?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`
  );
  if (!res.ok) {
    throw new Error("Failed to fetch my calendar");
  }
  return res.json();
}

// --- Time-off request API functions ---

export async function createTimeOffRequest(data: { date: string; notes?: string | null }): Promise<TimeOffRequest> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/time-off-request`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(data),
  });
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to create time-off request");
  }
  return res.json();
}

export async function fetchMyTimeOffRequests(): Promise<TimeOffRequestsResponse> {
  const res = await fetchWithAuth(`${API_BASE}${API_V1}/scheduler/time-off-requests`);
  if (!res.ok) {
    throw new Error("Failed to fetch time-off requests");
  }
  return res.json();
}

export async function approveTimeOffRequest(timeOffId: string): Promise<void> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/time-off-requests/${timeOffId}/approve`,
    { method: "PATCH" }
  );
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to approve time-off request");
  }
}

export async function declineTimeOffRequest(timeOffId: string): Promise<void> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/time-off-requests/${timeOffId}/decline`,
    { method: "PATCH" }
  );
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to decline time-off request");
  }
}

export async function deleteTimeOffRequest(timeOffId: string): Promise<void> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/time-off-requests/${timeOffId}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || "Failed to delete time-off request");
  }
}

export async function fetchTimeOffByDateRange(start: string, end: string): Promise<TimeOffRequestsResponse> {
  const res = await fetchWithAuth(
    `${API_BASE}${API_V1}/scheduler/time-off-requests?start_date=${encodeURIComponent(start)}&end_date=${encodeURIComponent(end)}`
  );
  if (!res.ok) {
    throw new Error("Failed to fetch time-off requests");
  }
  return res.json();
}
